### PR TITLE
NiPoPow proving mode impl

### PR DESCRIPTION
**Pull request by oskin1**
In this PR NiPoPow proving functionality added to the node.
Reference https://eprint.iacr.org/2017/963.pdf

Part of #211

### CODE REVIEW

File: src/main/scala/org/ergoplatform/modifiers/history/PoPowProofSuffix.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package org.ergoplatform.modifiers.history
+
+import org.ergoplatform.modifiers.ErgoPersistentModifier
+import org.ergoplatform.settings.Algos
+import scorex.core.ModifierTypeId
+import scorex.core.serialization.ScorexSerializer
+import scorex.util.ModifierId
+import scorex.util.serialization.{Reader, Writer}
+
+final case class PoPowProofSuffix(k: Int,
```
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/modifiers/history/PoPowProofSuffix.scala]: Done

File: src/main/scala/org/ergoplatform/modifiers/history/PoPowProofPrefix.scala [hunk 2]
```diff
@@ -0,0 +1,70 @@
+package org.ergoplatform.modifiers.history
+
+import org.ergoplatform.modifiers.ErgoPersistentModifier
+import org.ergoplatform.settings.{Algos, Constants}
+import scorex.core.ModifierTypeId
+import scorex.core.serialization.ScorexSerializer
+import scorex.util.serialization.{Reader, Writer}
+import scorex.util.{ModifierId, bytesToId, idToBytes}
+
+final case class PoPowProofPrefix(m: Int,
```
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/modifiers/history/PoPowProofPrefix.scala]: Done

File: src/main/scala/org/ergoplatform/modifiers/history/PoPowProof.scala [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package org.ergoplatform.modifiers.history
+
+import org.ergoplatform.modifiers.ErgoPersistentModifier
+import scorex.core.ModifierTypeId
+import scorex.core.serialization.ScorexSerializer
+import scorex.util.ModifierId
+import scorex.util.serialization.{Reader, Writer}
+
+final case class PoPowProof(prefix: PoPowProofPrefix, suffix: PoPowProofSuffix)
```
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/modifiers/history/PoPowProof.scala]: Done

File: src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala [hunk 4]
```diff
@@ -62,19 +66,65 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
...
+          // our requested list is half empty - request more missed modifiers
           self ! CheckModifiersToDownload
         } else {
           // headers chain is not synced yet, but our requested list is half empty - ask for more headers
           sendSync(statusTracker, h)
         }
       }
     }
+
+  override protected def processSync: Receive = {
```
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala]: Done

File: scalastyle-config.xml [hunk 5]
```diff
@@ -25,7 +25,7 @@
    <parameter name="regex"><![CDATA[[A-Z][A-Za-z]*]]></parameter>
   </parameters>
  </check>
- <check level="warning" class="org.scalastyle.scalariform.ObjectNamesChecker" enabled="true">
+ <check level="warning" class="org.scalastyle.scalariform.ObjectNamesChecker" enabled="false">
```
[the remark this answers is not in the document]
Comment (oskin1) [reply scalastyle-config.xml]: Setting rolled back.

File: src/main/scala/org/ergoplatform/modifiers/history/PoPowAlgos.scala [hunk 6]
```diff
@@ -43,14 +41,14 @@ object PoPowAlgos {
...
       Seq(prevHeader.id)
     }
-  }
 
   /**
     * Packs interlinks into extension key-value format.
     */
   @inline def packInterlinks(links: Seq[ModifierId]): Seq[(Array[Byte], Array[Byte])] = {
+    @scala.annotation.tailrec
     def loop(rem: List[(ModifierId, Int)],
```
[the remark this answers is not in the document]
Comment (oskin1) [reply src/main/scala/org/ergoplatform/modifiers/history/PoPowAlgos.scala]: I prefer using tail recursion as it is more readable and doesn't imply any mutable state.

**Comment by oskin1** [comment]
> I wonder what is the expected recursion level in bestArg method.
> It is written in extremely inefficient way: to calculate maximal element it first materialize sequence in memory using :+ operator. Constructing sequence of 100k integer pairs in such way takes at my desktop about 80 seconds. Also at each iteration chain is filtered using the following condition (absoluteLevelOf(_) >= level). What is the expected chain length? If it can be > 10k, then this operation will be also very slow. So it is possible to speedup this function several orders of magnitude if we calculate maximum inductively. But certainly it is critical only if size of chain and number of iterations are large enough.

NiPoPow chain length is `log2(chain.length)`. Please specify what do you mean by "calculate maximum inductively"

**Comment by knizhnik** [comment]
> NiPoPow chain length is log2(chain.length). Please specify what do you mean by "calculate maximum inductively"

I mean without materialization of intermediate sequence, something like this:
  def bestArg(chain: Seq[Header])(m: Int): Int = {
    @scala.annotation.tailrec
    def loop(max : Int, level: Int, acc: Seq[(Int, Int)] = Seq.empty): Int =
      val count = chain.count(absoluteLevelOf(_) >= level)
      if (count >= m) {
        max.max((math.pow(2, level) * count).toInt) // 2^µ * |C↑µ|
      } else {
	 max
      }
    loop(max = chain.size, level = 1) // Supposing each header is at least of level 0.
  }

**Comment by oskin1** [comment]
> > NiPoPow chain length is log2(chain.length). Please specify what do you mean by "calculate maximum inductively"
> 
> I mean without materialization of intermediate sequence, something like this:
> 
> ```
>   def bestArg(chain: Seq[Header])(m: Int): Int = {
>     @scala.annotation.tailrec
>     def loop(max : Int, level: Int): Int =
>       val count = chain.count(absoluteLevelOf(_) >= level)
>       if (count >= m) {
>         max.max((math.pow(2, level) * count).toInt) // 2^µ * |C↑µ|
>       } else {
> 	 max
>       }
>     loop(max = chain.size, level = 1) // Supposing each header is at least of level 0.
>   }
> ```

@knizhnik, I misunderstood you question a bit and hadn't actually answered -- looking at the code you provided I see that you have performance concerns about accumulator in recursive function which can not be greater than 2^8 in length (its length is supposed to be = max level).

**Comment by knizhnik** [comment]
> - looking at the code you provided I see that you have performance concerns about accumulator in recursive function which can not be greater than 2^8 in length (its length is supposed to be = max level).
@oskin1 
I have asked about possible maximal values of level and chain size.  There are three performance problems in the original version of the function:
1. Unneeded materialization of list of pairs (max can be calculated immediately)
2. Using copy operator :+ for constructing this list (instead of in-place append)
3. Coping of chain instead of just counting elements in it
Certainly if level of recursion is not create than 256, 1 and 2 should not cause a problem.
But I wonder what can be the maximal chain size? If we are specking about full blockchain, then as far as I understand it can be very large (now about 100k, but further can grow up to hundreds of millions). bestArg and many other functions introduced in this patch manipulates with header chains as if them are small enough  and it is not a big deal to copy or traverse it. Is it really correct assumption?

**Comment by oskin1** [comment]
> > * looking at the code you provided I see that you have performance concerns about accumulator in recursive function which can not be greater than 2^8 in length (its length is supposed to be = max level).
> 
> @oskin1 I have asked about possible maximal values of level and chain size. There are three performance problems in the original version of the function:
> 
> 1. Unneeded materialization of list of pairs (max can be calculated immediately)
> 2. Using copy operator :+ for constructing this list (instead of in-place append)
> 3. Coping of chain instead of just counting elements in it
> 
> Certainly if level of recursion is not create than 256, 1 and 2 should not cause a problem.
> But I wonder what can be the maximal chain size? If we are specking about full blockchain, then as far as I understand it can be very large (now about 100k, but further can grow up to hundreds of millions). bestArg and many other functions introduced in this patch manipulates with header chains as if them are small enough and it is not a big deal to copy or traverse it. Is it really correct assumption?

1. Materialized sequence can not contain more than 256 elements, so this doesn't make much difference in sense of performance while making code easier to read.
2. We are trying to avoid mutability in Scala when possible
3. Chain is not copied anywhere in this piece of code

**Comment by knizhnik** [comment]
@oskin1 
> Chain is not copied anywhere in this piece of code
Hmmm, filter() is creating copy of substantial part of sequence, isn't it:
`val args = chain.filter(absoluteLevelOf(_) >= level)` 
Similar operations with sequences are performed in many other methods:
```
val subChain = chain.dropRight(params.k)
chain.map(_.header), subChain.map(_.header),
subChain(subChain.size - params.m),
val suffix = chain.takeRight(params.k)
val superChainSuffixSize = superChain.takeRight(mValue).size
val downChainSuffixSize = downChain.takeRight(mValue).size
val downChain = chain.dropWhile(_ == superChain.head).takeWhile(_ == superChain.last) // C'↓
that.headersChain.filter(_.height > h.height))
```
So there are no places when chains are explicitly copied, but there are many places  where size of result is the same or comparable (depending on values) of parameters with size of input sequence.
Definitely this list is no complete and I expect that some operations like getting chain prefix/suffix do not require mass copying.

I am not familiar with Ergo system, so it is hard to me to predict where it may be critical and where not.  Also there is well known rule: do not try to optimize non stable code (may be there is  bottleneck at all or it is in other places). So may be my concerns are irrelevant. But if we are really going to support huge chains (I failed to get answer for the question what is the expected maximal chain length we are going to handle), then may be we should put more attention to such places and add tests which try to manipulate with chains with millions elements at VM with 3Gb memory limit?

File: src/test/resources/application.conf [hunk 7]
```diff
@@ -50,6 +47,12 @@ ergo {
 
     # Minimal fee amount of transactions mempool accepts
     minimalFeeAmount = 0
+
+    poPow {
+      # Download PoPoW proof on node bootstrap
```
Comment (kushti) [inline src/test/resources/application.conf]: This comment is misleading

File: src/main/scala/org/ergoplatform/nodeView/history/components/BlockSectionComponent.scala [hunk 8]
```diff
@@ -10,12 +10,13 @@ import scala.util.Try
   * Trait that declares interfaces for validation and processing of various
   * block sections: BlockTransactions, ADProofs, etc.
   */
-trait BlockSectionProcessor extends ScorexEncoding {
+trait BlockSectionComponent extends ScorexEncoding {
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/nodeView/history/components/BlockSectionComponent.scala]: While it is natural to call an entity which is doing processing and validation "Processor", generic "Component" does not say much about what could be inside it (Section & Component sound like tautology also). Why this renaming then?

File: src/main/scala/org/ergoplatform/modifiers/history/ADProofs.scala [hunk 9]
```diff
@@ -70,7 +70,7 @@ case class ADProofs(headerId: ModifierId,
 }
 
 object ADProofs extends ApiCodecs {
-  val modifierTypeId: ModifierTypeId = ModifierTypeId @@ (104: Byte)
+  val TypeId: ModifierTypeId = ModifierTypeId @@ (104: Byte)
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/modifiers/history/ADProofs.scala]: Why this renaming? new name is too generic

File: src/main/scala/org/ergoplatform/nodeView/history/components/Configuration.scala [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package org.ergoplatform.nodeView.history.components
+
+import org.ergoplatform.settings.ErgoSettings
+import scorex.core.utils.NetworkTimeProvider
+
+/**
+  * A component proving an access to configuration.
+  */
+trait Configuration {
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/nodeView/history/components/Configuration.scala]: Configuration for what?

**Comment by oskin1** [comment]
> @oskin1
> 
> > Chain is not copied anywhere in this piece of code
> 
> Hmmm, filter() is creating copy of substantial part of sequence, isn't it:
> `val args = chain.filter(absoluteLevelOf(_) >= level)`
> Similar operations with sequences are performed in many other methods:
> 
> ```
> val subChain = chain.dropRight(params.k)
> chain.map(_.header), subChain.map(_.header),
> subChain(subChain.size - params.m),
> val suffix = chain.takeRight(params.k)
> val superChainSuffixSize = superChain.takeRight(mValue).size
> val downChainSuffixSize = downChain.takeRight(mValue).size
> val downChain = chain.dropWhile(_ == superChain.head).takeWhile(_ == superChain.last) // C'↓
> that.headersChain.filter(_.height > h.height))
> ```
> 
> So there are no places where chains are explicitly copied, but there are many places where size of result is the same or comparable (depending on values of parameters) with size of input sequence.
> Definitely this list is no complete and I expect that some operations like getting chain prefix/suffix do not require mass copying.
> 
> I am not familiar with Ergo system, so it is hard to me to predict where it may be critical and where not. Also there is well known rule: do not try to optimize non stable code (may be there is bottleneck at all or it is in other places). So may be my concerns are irrelevant. But if we are really going to support huge chains (I failed to get answer for the question what is the expected maximal chain length we are going to handle), then may be we should put more attention to such places and add tests which try to manipulate with chains with millions elements at VM with 3Gb memory limit?

I agree that we should pay attention to such optimizations as chain grow linearly. As such problematic pieces of code are spread throughout the whole project it's worth performing such optimizations in separate PR with benchmarks for each such hotspot.

**Pull request closed without merging**